foundationdb/src/api.rs

```diff
@@ -116,214 +342,107 @@ pub struct NetworkBuilder {
+///     // Initialize FoundationDB normally, skipping the auto-stop functionality provided by `boot`
+///     foundationdb::boot().expect("failed to initialize FoundationDB");
+///
+///     // Perform operations with FoundationDB
+///
+///     // Dropping the `guard` ensures the network is stopped automatically when no other instances exist
+///     drop(guard);
+/// }
+/// ```
+pub struct NetworkAutoStop {
+    _private: (),
 }
 
-impl NetworkRunner {
-    /// Start the foundationDB network event loop in the current thread.
-    ///
-    /// # Safety
-    ///
-    /// This method is unsafe because you **MUST** call the `stop` method on the
-    /// associated `NetworkStop` before the program exit.
-    ///
-    /// This will only returns once the `stop` method on the associated `NetworkStop`
-    /// object is called or if the foundationDB event loop return an error.
-    pub unsafe fn run(self) -> FdbResult<()> {
-        self._run()
-    }
-
-    fn _run(self) -> FdbResult<()> {
-        {
-            let (lock, cvar) = &*self.cond;
-            let mut started = lock.lock().unwrap();
-            *started = true;
-            // We notify the condvar that the value has changed.
-            cvar.notify_one();
-        }
-
-        error::eval(unsafe { fdb_sys::fdb_run_network() })
-    }
-
-    unsafe fn spawn(self) -> thread::JoinHandle<()> {
-        thread::spawn(move || {
-            self.run().expect("failed to run network thread");
-        })
+impl Default for NetworkAutoStop {
+    fn default() -> Self {
+        Self::new()
     }
 }
 
-/// A condition object that can wait for the associated `NetworkRunner` to actually run.
-///
-/// Most of the time you should never need to use this directly and use `boot()`.
-pub struct NetworkWait {
-    cond: Arc<(Mutex<bool>, Condvar)>,
-}
+impl NetworkAutoStop {
+    pub fn new() -> Self {
+        NETWORK_STOP_HANDLER_COUNTER.fetch_add(1, Ordering::Acquire);
+        Self { _private: () }
+    }
 
-impl NetworkWait {
-    /// Wait for the associated `NetworkRunner` to actually run.
-    ///
-    /// # Panics
+    /// Attempts to stop the FoundationDB network if it is running.
     ///
-    /// Panics if the internal lock cannot is poisoned
-    pub fn wait(self) -> NetworkStop {
-        // Wait for the thread to start up.
-        {
-            let (lock, cvar) = &*self.cond;
-            let mut started = lock.lock().unwrap();
-            while !*started {
-                started = cvar.wait(started).unwrap();
-            }
+    /// This method decreases the internal reference counter tracking active network users. If the counter
+    /// reaches zero and the network thread is still running, it invokes [`stop_network`] to shut down the
+    /// FoundationDB network cleanly.
+    pub fn stop_network(&self) -> Result<(), FdbError> {
+        let previous_count = NETWORK_STOP_HANDLER_COUNTER.fetch_sub(1, Ordering::Acquire);
+        if previous_count == 1 && is_network_thread_running() {
```

Comment:
race condition with boot()
public so a user can manually decrement the counter...